jgrapht-core/src/main/java/org/jgrapht/alg/tour/WarnsdorffRuleKnightTourHeuristic.java

```diff
@@ -0,0 +1,810 @@
+ *
+ * 1). $(1, 0) \to (0, 2)$ - denoted as $1$ on the picture below.
+ * 2). $(2, 0) \to (0, 1)$ - denoted as $2$ on the picture below.
+ * 3). $(n - 3, 0) \to (n - 1, 1)$ - denoted as $3$ on the picture below.
+ * 4). $(n - 2, 0) \to (n - 1, 2)$ - denoted as $4$ on the picture below.
+ * 5). $(0, m - 3) \to (1, m - 1)$ - denoted as $5$ on the picture below.
+ * 6). $(0, m - 2) \to (2, m - 1)$ - denoted as $6$ on the picture below.
+ * 7). $(n - 3, m - 1) \to (n - 1, m - 2)$ - denoted as $7$ on the picture below.
+ * 8). $(n - 2, m - 1) \to (n - 1, m - 3)$ - denoted as $8$ on the picture below.
+ *
+ * #########################################
+ * #*12*********************************34*#
+ * #2*************************************3#
+ * #1*************************************4#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #***************************************#
+ * #6*************************************8#
+ * #5*************************************7#
+ * #*65*********************************78*#
+ * #########################################
+ *
+ * If you are confused with the formal definition of the structured knight's tour
+ * please refer to illustration on the page $3$ of the paper
+ * "An efficient algorithm for the Knight’s tour problem " by Ian Parberry.
+ *
+ * One more feature of this implementation is that it provides an option to return a shifted knight's tour,
+ * where all cell's coordinates are shifted by some values.
+ * Basically it is the same as knight's tour of some piece of the board.
+ */
+
+public class WarnsdorffRuleKnightTourHeuristic {
+
+    /**
+     * Width of the board.
+     */
+
+    private int n;
+
+    /**
+     * Height of the board.
+     */
+
+    private int m;
+
+    /**
+     * 2d array that stores information whether or not the cell has been visited.
+     */
+
+    private boolean[][] chessBoard;
+
+    /**
+     * Auxiliary array for offset in x coordinate when performing a move.
+     */
+
+    private final static int[] DX = new int[]{1, 2, 2, 1, -1, -2, -2, -1};
+
+    /**
+     * Auxiliary array for offset in y coordinate when performing a move.
+     */
+
+    private final static int[] DY = new int[]{2, 1, -1, -2, -2, -1, 1, 2};
+
+    /**
+     * Constructor.
+     * @param n width and height of the board.
+     */
+
+    public WarnsdorffRuleKnightTourHeuristic(int n) {
+        if (n < 3) {
+            throw new UnsupportedOperationException("You set incorrect board size!");
```

Comment:
Use `IllegalArgumentExeption` and don't make the message so personal; msg = "Incorrect board size: " + n would be better.